src/util/recentlyEditedTree.js

```diff
@@ -0,0 +1,214 @@
+      }
+    }
+  }
+}
+
+/**
+   * Deletes node from existing tree object using mutation.
+   * @param {object} tree nested object representing tree
+   * @param {string[]} oldPath array of string representing path
+   * @param {string[]} newPath array of string representing path
+   * @param {boolean} [timestampUpdate=true] if false it doesn't update lastUpdated property of all affected leaf nodes (used by nodeMove)
+   */
+const nodeDelete = (tree, oldPath, timestampUpdate = true) => {
+  const oldContext = pathToContext(oldPath)
+  const { node: commonNode, path: commonPath } = findTreeDeepestSubcontext(tree, oldContext)
+  if (commonNode) {
+    if (oldContext.length > commonPath.length) return // if node is not already available in the tree
+    const closestAncestor = findClosestAncestorWithMultipleChildren(tree, commonPath)
+    const nodeToMergeIntoPath = commonPath.length - closestAncestor.path.length <= 2 ? closestAncestor.path : false
+    if (nodeToMergeIntoPath) {
+      // checking all the nodes from the direct parent or grandparent , if available updating all the descendants lastUpdated
+      const pathBeingMerged = commonPath.slice(0, closestAncestor.path.length + 1)
+      _.unset(tree, pathBeingMerged) // deleting the merged path
+      findTreeDescendants(tree, nodeToMergeIntoPath).forEach(descendant => {
+        _.set(tree, pathToContext(descendant.path), { leaf: true, lastUpdated: timestampUpdate ? timestamp() : descendant.lastUpdated, path: descendant.path })
+      })
+    }
+    // if cannot be merged to either parent or grandParent then just making parent path as the leaf node *** exception if parent path is __ROOT__ node ***
+    else {
+      _.unset(tree, commonPath)
+      _.set(tree, contextOf(commonPath), { leaf: true, lastUpdated: timestamp(), path: oldPath.slice(0, contextOf(commonPath).length) })
+    }
+  }
+}
+
+/**
+   * Moves a node between branches in the existing tree object using mutation.
+   * @param {object} tree nested object representing tree
+   * @param {string[]} oldPath array of string representing path
+   * @param {string[]} newPath array of string representing path
+   */
+const nodeMove = (tree, oldPath, newPath) => {
+  const oldContext = pathToContext(oldPath)
+  const newContext = pathToContext(newPath)
+
+  const { node: oldNode, path: commonOldPath } = findTreeDeepestSubcontext(tree, oldContext)
+  const { path: commonNewPath } = findTreeDeepestSubcontext(tree, newContext)
+
+  if (oldNode && commonOldPath.length === oldContext.length) {
+    // if node is already available
+    if (equalArrays(commonOldPath, commonNewPath)) {
+      // if moved with in same place (only rank has changed) (to prevent traversing descendants)
+      nodeChange(tree, oldPath, newPath)
+    }
+    else {
+      if (oldNode.leaf) {
+        nodeChange(tree, contextOf(newPath).concat({ value: '' }), newPath)
```

Comment:
Sure. If `oldContext` is already available node on the recently edited tree `if(equalArrays(commonPath, oldContext))` then `nodeChange` will only update it's descendants. When we move nodes in the tree we delete `oldNode` and add it to destination branch. So we add a empty node to the `contextOf` new path to add it as new node to the new branch . 

For example:

A.B.C.D.E.F
A.B.C.G.H.I

If we want to add A.B.C.K.L , then ['A','B','C',''] will be `oldPath` and  ['A','B','C','K','L'] will be `newPath`.

Reply:
Okay, thanks for explaining! Can we add an explicit `nodeAdd` method? It's not ideal to have a magic value like `{ value: '' }` that is used to activate a particular code path. I see too realistic options:

1. Hide the magic value in a `nodeAdd` method that delegates to `nodeChange` using `{ value: DUMMY_TOKEN }` to make it explicit. Add a comment to explain how the unconventional use of `nodeChange` achieves the desired effect.
2. Write separate `nodeAdd` and `nodeChange` methods, and abstract the common logic (may involve unnecessary effort at this point.)

Reply:
The problem with this approach is that we are using `existingThoughtChange` and when new empty thought is added and first value is entered we get same pattern of `oldPath` and `newPath` by `existingThoughtChange`.

For example:
If new empty `subthought` is added in A.B.C.D and we write value E inside the new empty thought. We get `oldPath` ['A','B','C','D',''] and `newPath` ['A','B','C','D','E']. 

What I mean to say is that we don't know if a new thought has been added to the tree because recently edited thoughts methods are called only on `existingThoughtChange`. Also we add nodes to the recently edited tree only after `nodeChange` doesn't find already available nodes. If it finds one it updates it's descendants **(change)** and if node is not available **(possible new node)** it makes sure that there are no merge able relations between already available nodes before adding a new leaf to the branch. So due to the fact that we don't know if new path will be added to the recently edited tree as leaf node, we have to first check for `nodeChange` and only then we can perform addition of new leaf node to the tree.

In case of node move we have to create a `oldPath` and a `newPath` for moved branches. Since `nodeChange` works as I explained above example. We have to provide same pattern of `oldPath` and `newPath` . So we have to add `{value:''}`. 

I think we should go with your first suggestion and write a function `nodeAdd` that adds `{ value: DUMMY_TOKEN }` to the provided `oldPath` and explain the logic there. `nodeAdd` will be used where we have to imitate new thought being add.